### PR TITLE
#30: Allow using an existing customerId (closes #30)

### LINKED ISSUE #30: Allow using an existing customerId
[Trello card](https://trello.com/c/59148a081c332fb44d56a4b7/)

## requirements

The library should support use cases where the ContactHub customerId is already known.

1. Allow users of the library to provide a `customerId` as part of a `ch('customer', customerObject)` call. `customerObject` should accept an optional key with name `id` that can contain the `customerId`.

2. Allow users of the library to provide a `customerId` using a query param named `clabId`. Its effect should be the same than the `id` key described above.

## specs

A `customerId` can be provided via a `id` key in `customerObject`, or via a `clabId` key in the query string.

### query string

If the `clabId` key is set in the query string, perform this implicit call as part of the bootstrap in the `config` method: `ch('customer', { id: clabId })`

### customer method

* If a customerId is not provided, maintain current behaviour
* If a customerId is provided but no other customer data (this can be achieved using the `clabId` query param or by calling `ch('customer', { id: 'some-id' })`)
  * If the same customerId is also stored in the cookie, do nothing
  * If a different customerId is stored in the cookie, do nothing (a different user could be logged in)
  * If no customerId is currently stored in the cookie, try reconciling the current sessionId with the provided customerId
    * If the reconciliation is successful (API returns 201), update the cookie with the new customerId
    * If the reconciliation fails because the provided customerId does not exist (API returns 404), do nothing and return
* If a customerId is provided along with some customer data (`externalId`, `base`, `extended`, `extra` or `tags`)
  * If the same customerId is also stored in the cookie, maintain current behaviour (perform a customer update (PATCH), then update the cookie and return)
  * If a different customerId is stored in the cookie, reset cookie with a new sessionId, then proceed as if no customerId was stored in the cookie (see below)
  * If no customerId is currently stored in the cookie, try reconciling the current sessionId with the provided customerId
    * If the reconciliation is successful (API returns 201), perform a customer update (PATCH), then update the cookie and return
    * If the reconciliation fails because the provided customerId does not exist (API returns 404), do nothing and return

## misc
* The case where "a different customerId is stored in the cookie" is handled differently based on whether some additional customer data is sent along. The reason for this is that if we receive some personal information about the customer, we have more confidence that this information is consistent. If we only receive an id and the cookie contains a different one, we play safe and assume the id contained in the cookie is more accurate.
* The reconciliation can also fail if sessionId and customerId are already associated (the API returns `400` with message `Session value [...] already exist`) -- this should never happen, because if the customerId and the sessionId are already associated, the customerId should also be in the cookie. If it does happen for any reason, it should be ok to fall into the `404` case above and do nothing.

### PR DESCRIPTION
**Pull request by dhinus**
Closes #30

### tests performed
* added some new unit tests, no regression on the existing ones
* tested manually with `npm run example`

### tests not performed (domain coverage)
I have not tested all the possible edge cases (e.g. sending different IDs from the query string and the `customer` method). I expect this feature to be used by a minority of users, and it seems to work correctly for the scenario described in the requirements.

### CODE REVIEW

File: test/customer.js [hunk 1]
```diff
@@ -64,145 +64,270 @@ describe('Customer API:', () => {
...
+            done();
+          });
+        });
+      });
+
+      it('stores a hash of the customer data for future calls', (done) => {
+        whenDone(() => {
+          requests[0].respond(200, {}, JSON.stringify({ id: 'new-cid' }));
+          whenDone(() => {
+            expect(getCookie().hash).not.to.be.undefined;
```
Comment (dhinus) [inline test/customer.js]: These lines were not changed in this PR but only moved. I'll leave them as they are for now to minimise the changes.

File: test/customer.js [hunk 2]
```diff
@@ -64,145 +64,270 @@ describe('Customer API:', () => {
...
+        });
+      });
     });
 
-    it('creates a new customer', (done) => {
-      whenDone(() => {
+    describe('and a customerId is stored in the cookie', () => {
+      beforeEach(() => {
+        setConfig();
+        cookies.set(cookieName, Object.assign(getCookie(), {
```
Comment (dhinus) [inline test/customer.js]: `store` is just one of many different calls of `cookies.set`. It is true that it is reused in a few places, but there are other places where `cookies.set` is called with different parameters. `cookies.set` already comes from an external library so I think it provides a reasonable level of abstraction.

I'm up for discussing a refactor of all cookie-related lines in a separate issue/PR.
